src/react/components/VisualFlowEditor.js

```diff
@@ -0,0 +1,107 @@
+/*
+  This file is part of Astarte.
+
+  Copyright 2020 Ispirata Srl
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+
+import React, { useCallback, useEffect, useState } from "react";
+import { CanvasWidget } from '@projectstorm/react-canvas-core';
+import createEngine, { DiagramModel } from '@projectstorm/react-diagrams';
+import { NativeBlockFactory } from './NativeBlock.js';
+
+const filterSortBlocks = (blocks, type) => {
+  if (!blocks || blocks.length === 0) {
+    return [];
+  }
+
+  return blocks
+    .filter((block) => block.type === type)
+    .sort((block1, block2) => block1.name > block2.name);
+};
+
+const BlockMenuItem = ({block}) => (
+  <div
+    className={`block-item ${block.type}`}
+    onDragStart={(e) => e.dataTransfer.setData("block-name", block.name)}
+    draggable
+  >
+    {block.name}
+  </div>
+);
+
+const EditorSidebar = ({blocks}) => (
+  <div className="flow-editor-sidebar">
+    <div className="block-label">Producer</div>
+    {filterSortBlocks(blocks, 'producer').map((block) => (
+      <BlockMenuItem key={block.name} block={block} />
+    ))}
+    <div className="block-label">Producer & consumer</div>
+    {filterSortBlocks(blocks, 'producer_consumer').map((block) => (
+      <BlockMenuItem key={block.name} block={block} />
+    ))}
+    <div className="block-label">Consumer</div>
+    {filterSortBlocks(blocks, 'consumer').map((block) => (
+      <BlockMenuItem key={block.name} block={block} />
+    ))}
+  </div>
+);
+
+function getEngine(model, nodeFactory) {
+  const engine = createEngine();
+  engine.getNodeFactories().registerFactory(nodeFactory);
+  engine.setModel(model);
+
+  return engine;
+}
+
+const VisualFlowEditor = ({className = "", blocks, model, nodeFactory, onNodeSettingsClick}) => {
+  const [engine, setEngine] = useState(getEngine(model, nodeFactory));
+
+  const addBlock = useCallback((name, position) => {
```

Comment:
Can we have `model` and `nodeFactory` generated on component's mount instead on delegating the work to the parent component and relying on props?